File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    idNameDict = {}
    message = []
    for r in record:
        command = r.split(' ')
        action = command[0]
        id = command[1]
        if(action == 'Enter'):
            idNameDict[id] = command[2]
            message.append({'action': action, 'id': id})
        elif(action == 'Leave'):
            message.append({'action': action, 'id': id})
        elif(action == 'Change'):
            idNameDict[id] = command[2]
        
    result = []
    for m in message:
        name = idNameDict[m['id']]
        if m['action']=='Enter':
            result.append(f'{name}님이 들어왔습니다.')
        elif m['action']=='Leave':
            result.append(f'{name}님이 나갔습니다.')
            
    return result

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String, String> user = new HashMap<>();
        for(String r : record) {
            String[] split = r.split(" ");
            String oper = split[0];
            if(oper.equals("Leave")) continue;
            
            String id = split[1];
            String name = split[2];
            
            user.put(id, name);
        }
        
        List<String> list = new ArrayList<>();
        
        for(String r : record) {
            String[] split = r.split(" ");
            String oper = split[0];
            String id = split[1];
            
            if(oper.equals("Enter")) {
                list.add(user.get(id)+"님이 들어왔습니다.");
            } else if(oper.equals("Leave")) {
                list.add(user.get(id)+"님이 나갔습니다.");
            }
        }
        String[] answer = new String[list.size()];
        for(int i=0; i<list.size(); i++) {
            answer[i] = list.get(i);
        }
        return answer;
    }
}